pkg/test/framework/api/components/apps.go

```diff
@@ -43,35 +43,30 @@ type Apps interface {
 	GetAppOrFail(name string, t testing.TB) App
 }
 
+// Service represents a deployed service within the mesh.
+type Service interface {
+	ClusterIP() string
+}
+
 // App represents a deployed fake App within the mesh.
 type App interface {
 	Name() string
 	Endpoints() []AppEndpoint
 	EndpointsForProtocol(protocol model.Protocol) []AppEndpoint
 	Call(e AppEndpoint, opts AppCallOptions) ([]*echo.ParsedResponse, error)
 	CallOrFail(e AppEndpoint, opts AppCallOptions, t testing.TB) []*echo.ParsedResponse
+	Service() Service
 }
 
 // AppCallOptions defines options for calling a DeployedAppEndpoint.
 type AppCallOptions struct {
-	// Secure indicates whether a secure connection should be established to the endpoint.
-	Secure bool
-
-	// Protocol indicates the protocol to be used.
-	Protocol AppProtocol
-
-	// UseShortHostname indicates whether shortened hostnames should be used. This may be ignored by the environment.
-	UseShortHostname bool
-
 	// Count indicates the number of exchanges that should be made with the service endpoint. If not set (i.e. 0), defaults to 1.
 	Count int
-
-	// Headers indicates headers that should be sent in the request. Ingnored for WebSocket calls.
-	Headers http.Header
 }
 
 // AppEndpoint represents a single endpoint in a DeployedApp.
 type AppEndpoint interface {
+	URL() *url.URL
```

Comment:
this shouldn't be in the API since doesn't make sense for all platforms

Reply:
For me a URL should make sense on all platforms, even in a native environment.

If you look at the [native app implementation](https://github.com/istio/istio/blob/release-1.1/pkg/test/framework/runtime/components/apps/native.go#L432) in release-1.1 you will also find a function, which builds a URL. Why should we not expose this URL to the outside of the test framework?

Reply:
Same reasoning as exposing the ClusterIP above. The app API should provide all the things we need ... exposing this just affords hackery which may not work the same way on every environment.